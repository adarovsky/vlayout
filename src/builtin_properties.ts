import { TypeDefinition } from './types';
import { EMPTY, Observable } from 'rxjs';
import { Expression } from './expression';

export class BuiltinProperty extends Expression {
    constructor(readonly keyPath: string, readonly typeDefinition: TypeDefinition, readonly sink: Observable<any>) {
        super(0, 0);
    }

    instantiate(): this {
        const v = new (this.constructor as typeof BuiltinProperty)(this.keyPath, this.typeDefinition, this.sink);
        return v as this;
    }

    toString(): string {
        return this.keyPath;
    }
}
