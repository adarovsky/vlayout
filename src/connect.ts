import React, { createElement } from 'react';
import { combineLatest, Observable, of, Subscription } from 'rxjs';
import { fromPairs, toPairs, unzip, zip } from 'lodash';

export type ObservableMapObject<S = any> = {
    [K in keyof S]: Observable<S[K]>|S[K]
}

export function connect<P, S = {}>(WrappedComponent: React.FC<P> | React.ComponentClass<P, S>) {
    return class Connector extends React.PureComponent<ObservableMapObject<P>, P> {
        subscription = new Subscription();

        componentDidMount(): void {
            this.subscription.unsubscribe();
            const pairs = toPairs(this.props);
            const [keys, values] = unzip(pairs);
            this.subscription = combineLatest(values.map(v => v instanceof Observable ? v : of(v))).subscribe(s => {
                const pairs = zip(keys, s);
                const state = fromPairs(pairs);
                this.setState(state as P);
            });
        }

        componentWillUnmount(): void {
            this.subscription.unsubscribe();
        }

        render(): React.ReactElement<any, string | React.JSXElementConstructor<any>> | string | number | {} | React.ReactNodeArray | React.ReactPortal | boolean | null | undefined {
            if (this.state) {
                return createElement(WrappedComponent, this.state);
            } else {
                return null;
            }
        }
    }
}
