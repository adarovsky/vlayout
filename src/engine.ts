import { Inputs } from './inputs';
import {
    Dictionary,
    EnumDefinition,
    ListDefinition,
    ListDefinitionItem,
    TypeDefinition,
    Types,
} from './types';
import { EnumValue, Expression } from './expression';
import {
    Ceil,
    ColorAlpha,
    Floor,
    FontFamily,
    FontSized,
    FontTyped,
    FunctionImplementationI,
    Image,
    LocalizedNumber,
    LocalizedString,
    Max,
    Min,
    Round,
    ShortLocalizedNumber,
    StringPrefix,
    StringSuffix,
    SubString,
    ToCapitalCase,
    ToLower,
    ToUpper,
} from './builtin_functions';
import React from 'react';
import { ViewListReference, ViewReference } from './view_reference';
import { View } from './view';
import { ReactViewProps } from './react_views';
import { Button, TextField } from './primitives';
import { Observable } from 'rxjs';
import {
    ListClickHandler,
    ListEnterHandler,
    ListModelItem,
    ListTextChangeHandler,
} from './list';
import { isEqual } from 'lodash';
import invariant from 'ts-invariant';

export class Engine {
    readonly inputs = new Inputs(this);
    readonly types = new Types(this);
    readonly functions: FunctionImplementationI[];
    private readonly referencedViews: Dictionary<ViewReference> = {};
    private readonly buttons: Dictionary<() => Promise<void>> = {};
    private readonly textFields: Dictionary<{
        onChange: (s: string) => void;
        onEnter: () => void;
    }> = {};
    private readonly listButtons: Dictionary<ListClickHandler> = {};
    private readonly listTextFields: Dictionary<{
        onChange: ListTextChangeHandler;
        onEnter: ListEnterHandler;
    }> = {};
    private readonly referencedListViews: Dictionary<ViewListReference> = {};
    readonly valueSnapshot: {
        inputs: Dictionary<any>;
        properties: Dictionary<any>;
    } = {
        inputs: {},
        properties: {},
    };

    constructor(readonly debug = false, readonly verboseIds: string[] = []) {
        this.functions = [
            new LocalizedNumber(this),
            new ShortLocalizedNumber(this),
            new Image(this),
            new FontFamily(this),
            new FontTyped(this),
            new FontSized(this),
            new ColorAlpha(this),
            new StringPrefix(this),
            new StringSuffix(this),
            new SubString(this),
            new ToUpper(this),
            new ToLower(this),
            new ToCapitalCase(this),
            new Round(this),
            new Floor(this),
            new Ceil(this),
        ];
    }

    type(name: string): TypeDefinition | null {
        return this.types.type(name);
    }

    boolType(): TypeDefinition {
        return this.type('Bool')!;
    }

    numberType(): TypeDefinition {
        return this.type('Number')!;
    }

    colorType(): TypeDefinition {
        return this.type('Color')!;
    }

    stringType(): TypeDefinition {
        return this.type('String')!;
    }

    fontType(): TypeDefinition {
        return this.type('Font')!;
    }

    imageType(): TypeDefinition {
        return this.type('Image')!;
    }

    variableForKeyPath(keyPath: string): Expression | null {
        const kp = keyPath.split('.');
        if (kp.length === 2) {
            const enumType = this.type(kp[0]);
            if (enumType instanceof EnumDefinition) {
                const v = new EnumValue(kp[1], 0, 0);
                v.linkEnum(enumType);
                return v;
            }
        }

        return this.inputs.input(keyPath, 0, 0);
    }

    functionFor(
        name: string,
        parameters: TypeDefinition[]
    ): FunctionImplementationI {
        if (name === '@') {
            for (let p of parameters) {
                if (p !== this.stringType()) {
                    throw new Error(`localized string receives only strings`);
                }
            }

            return new LocalizedString(this);
        } else if (name === 'Min') {
            for (let p of parameters) {
                if (p !== this.numberType()) {
                    throw new Error(`Min accepts only numbers`);
                }
            }
            if (parameters.length == 0) {
                throw new Error(`Not enough arguments to Min`);
            }

            return new Min(this);
        } else if (name === 'Max') {
            for (let p of parameters) {
                if (p !== this.numberType()) {
                    throw new Error(`Max accepts only numbers`);
                }
            }

            if (parameters.length == 0) {
                throw new Error(`Not enough arguments to Max`);
            }

            return new Max(this);
        } else {
            for (let f of this.functions) {
                if (f.name === name && isEqual(f.parameterTypes, parameters)) {
                    return f;
                }
            }

            throw new Error(
                `function ${name}(${parameters
                    .map((p) => p.toString())
                    .join(', ')}) not found`
            );
        }
    }

    functionsLoose(
        name: string,
        parametersCount: number
    ): FunctionImplementationI[] {
        return this.functions.filter(
            (f) =>
                f.name === name && f.parameterTypes.length === parametersCount
        );
    }

    registerView(
        key: string,
        createComponent: (parent: View) => React.ReactElement<ReactViewProps>
    ) {
        invariant(
            typeof createComponent === 'function',
            `createComponent for ${key} is empty`
        );
        this.referencedViews[key] = new ViewReference(createComponent);
    }

    registerInput(
        name: string,
        type: TypeDefinition,
        sink: Observable<any>
    ): void {
        invariant(sink, `input sink for ${name} is empty`);
        this.inputs.registerInput(name, type, sink);
    }

    registerButton(key: string, onClick: () => Promise<void>) {
        this.buttons[key] = onClick;
    }

    registerTextField(
        key: string,
        onChange: (s: string) => void,
        sink: Observable<string>,
        onEnter: () => void = () => {}
    ) {
        this.textFields[key] = { onChange, onEnter };
        this.inputs.registerInput(key, this.stringType(), sink);
    }

    registerEnum(name: string, values: Dictionary<any>): void {
        this.types.registerEnum(new EnumDefinition(this, name, values));
    }

    registerStandardEnum<EnumType>(
        name: string,
        e: StandardEnum<EnumType>
    ): void {
        this.types.registerEnum(
            new EnumDefinition(this, name, materializeEnum(e))
        );
    }

    registerList(name: string, fields: Dictionary<ListDefinitionItem>): void {
        this.types.registerList(new ListDefinition(this, name, fields));
    }

    registerListButton(
        name: string,
        onClick: (item: ListModelItem) => Promise<void>
    ) {
        this.listButtons[name] = onClick;
    }

    registerListTextField(
        name: string,
        onChange: ListTextChangeHandler,
        onEnter: ListEnterHandler = () => {}
    ) {
        this.listTextFields[name] = { onChange, onEnter };
    }

    registerListView(
        name: string,
        createComponent: (
            parent: View,
            modelItem: Observable<ListModelItem>
        ) => React.ReactElement<ReactViewProps>
    ) {
        this.referencedListViews[name] = new ViewListReference(createComponent);
    }

    viewForKey(key: string): View | null {
        if (this.buttons[key]) return new Button(this.buttons[key]);
        else if (this.textFields[key])
            return new TextField(
                key,
                this.textFields[key].onChange,
                this.textFields[key].onEnter
            );
        else if (this.referencedViews[key])
            return new ViewReference(this.referencedViews[key].createComponent);
        else return null;
    }

    listButtonForKey(key: string): ListClickHandler | null {
        return this.listButtons[key] || null;
    }

    listTextFieldForKey(
        key: string
    ): { onChange: ListTextChangeHandler; onEnter: ListEnterHandler } | null {
        return this.listTextFields[key] || null;
    }

    listViewForKey(key: string): ViewListReference | null {
        const view = this.referencedListViews[key];
        if (view) {
            return new ViewListReference(view.createComponent);
        }

        return null;
    }

    logInputValue(name: string, value: any): void {
        this.valueSnapshot.inputs[name] = value;
    }

    beginUpdates() {
        this.inputs.beginUpdates();
    }

    endUpdates() {
        this.inputs.endUpdates();
    }
}

export type StandardEnum<T> = {
    [id: string]: T | string;
    [nu: number]: string;
};

function camelize(str: string) {
    return str
        .replace(/(?:^\w|[A-Z]|\b\w|-|\s+)/g, function (match, index) {
            if (+match === 0) return ''; // or if (/\s+/.test(match)) for white spaces
            return index === 0 ? match.toLowerCase() : match.toUpperCase();
        })
        .replace('-', '');
}

function materializeEnum<EnumType>(e: StandardEnum<EnumType>) {
    return Object.keys(e)
        .filter((x) => isNaN(+x))
        .reduce((prev, cur) => ({ ...prev, [camelize(cur)]: e[cur] }), {});
}
