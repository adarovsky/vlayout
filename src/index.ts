export {Engine} from "./engine";
export {Layout} from "./layout";
export {ReactView, ReactViewState} from "./react_views";
import './vlayout.css';
