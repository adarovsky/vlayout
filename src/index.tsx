export {
    resizeObserver,
    pauseObserving,
    resumeObserving,
} from './resize_sensor';
export { Engine } from './engine';
export { Layout } from './layout';
import './vlayout.css';

export { ReactViewReference as ReactView } from './react_views';
export type { ReactViewProps, ReactViewState } from './react_views';
export { View } from './view';
export { ColorContainer as Color, ImageContainer as Image } from './types';
export type {
    SimpleListDefinitionItem,
    ListDefinitionItem,
    ComplexDefinitionItem,
} from './types';
export type { ListModelItem } from './list';
export { LayoutComponent } from './layout_component';
export { connect } from './connect';
export type { ObservableMapObject } from './connect';
