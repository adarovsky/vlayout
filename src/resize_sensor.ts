import {BehaviorSubject, EMPTY, Observable} from "rxjs";
import {distinctUntilChanged, finalize, shareReplay, switchMap} from "rxjs/operators";
import _ from "lodash";

const observers: [Element, Observable<ElementSize>][] = [];
export {observers};

export interface ElementSize {
    width: number;
    height: number;
}

let pausedCount = 0;
export function pauseObserving() {
    paused.next(++pausedCount > 0);
}

export function resumeObserving() {
    paused.next(Math.max(--pausedCount, 0) > 0);
}

const paused = new BehaviorSubject<boolean>(false);

export function resizeObserver(element: HTMLDivElement): Observable<ElementSize> {
    const obj = observers.find( o => o[0] === element);
    if (obj) {
        return obj[1];
    }

    const innerObserver = new Observable<ElementSize>(subscriber => {
        const style = getComputedStyle(element);
        const zIndex = ((style && style.zIndex) ? parseInt(style.zIndex)! : 0) - 1;

        let expand = document.createElement('div');
        expand.style.position = "absolute";
        expand.style.left = "0px";
        expand.style.top = "0px";
        expand.style.right = "0px";
        expand.style.bottom = "0px";
        expand.style.overflow = "hidden";
        expand.style.zIndex = `${zIndex}`;
        expand.style.visibility = "hidden";

        let expandChild = document.createElement('div');
        expandChild.style.position = "absolute";
        expandChild.style.left = "0px";
        expandChild.style.top = "0px";
        expandChild.style.width = "10000000px";
        expandChild.style.height = "10000000px";
        expand.appendChild(expandChild);

        let shrink = document.createElement('div');
        shrink.style.position = "absolute";
        shrink.style.left = "0px";
        shrink.style.top = "0px";
        shrink.style.right = "0px";
        shrink.style.bottom = "0px";
        shrink.style.overflow = "hidden";
        shrink.style.zIndex = `${zIndex}`;
        shrink.style.visibility = "hidden";

        let shrinkChild = document.createElement('div');
        shrinkChild.style.position = "absolute";
        shrinkChild.style.left = "0px";
        shrinkChild.style.top = "0px";
        shrinkChild.style.width = "200%";
        shrinkChild.style.height = "200%";
        shrink.appendChild(shrinkChild);

        element.appendChild(expand);
        element.appendChild(shrink);

        function setScroll() {
            expand.scrollLeft = 10000000;
            expand.scrollTop = 10000000;

            shrink.scrollLeft = 10000000;
            shrink.scrollTop = 10000000;
        }
        setScroll();

        let size = element.getBoundingClientRect();

        let currentWidth = size.width;
        let currentHeight = size.height;
        subscriber.next({width: currentWidth, height: currentHeight});

        let onScroll = function () {
            let size = element.getBoundingClientRect();

            let newWidth = size.width;
            let newHeight = size.height;

            if (newWidth !==currentWidth || newHeight !==currentHeight) {
                currentWidth = newWidth;
                currentHeight = newHeight;

                subscriber.next({width: currentWidth, height: currentHeight});
            }

            setScroll();
        };

        expand.addEventListener('scroll', onScroll);
        shrink.addEventListener('scroll', onScroll);

        return () => {
            expand.removeEventListener('scroll', onScroll);
            shrink.removeEventListener('scroll', onScroll);

            expand.remove();
            expandChild.remove();
            shrink.remove();
            shrinkChild.remove();
        }
    }).pipe(distinctUntilChanged(_.isEqual), shareReplay({bufferSize: 1, refCount: true}));

    const observer = paused.pipe(
        distinctUntilChanged(),
        switchMap(paused => paused ? EMPTY : innerObserver),
        finalize(() => {
            const index = observers.findIndex( o => o[0] === element);
            if (index >= 0) observers.splice(index, 1);
        })
    );
    observers.push([element, observer]);
    return observer;
}
