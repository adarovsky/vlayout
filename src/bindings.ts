import { Engine } from 'engine';
import { View } from 'view';
import { Button, TextField } from 'primitives';

export class Bindings {
    constructor(public readonly engine: Engine) {

    }

    registerSlot(name: string, type: string): void {
        if (type === 'listButton' && this.engine.listButtonForKey(name))
            return;

        if (type === 'listTextField' && this.engine.listTextFieldForKey(name))
            return;

        if (type === 'listView' && this.engine.listViewForKey(name))
            return;

        const v = this.viewForKey(name);
        if (!v) {
            throw new Error(`view ${name} is not registered`);
        }

        if (type === 'button' && !(v instanceof Button)) {
            throw new Error(`view ${name} is not a button, but ${v.viewType()}`);
        }

        if (type === 'textField' && !(v instanceof TextField)) {
            throw new Error(`view ${name} is not a text field, but ${v.viewType()}`);
        }
    }

    viewForKey(key: string): View | null {
        return this.engine.viewForKey(key);
    }

    toString(): string {
        return 'bindings {\n}\n';
    }
}
