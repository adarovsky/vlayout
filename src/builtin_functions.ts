import {Engine} from "./engine";
import {ColorContainer, FontContainer, ImageContainer, TypeDefinition} from "./types";
import {combineLatest, Observable} from "rxjs";
import {map} from "rxjs/operators";

export interface FunctionImplementationI {
    returnType: TypeDefinition;
    name: string;
    parameterTypes: TypeDefinition[];

    sink(parameters: Observable<any>[]): Observable<any>;
}

export class FunctionImplementation {
    constructor(public readonly engine: Engine) {
    }
}

export class LocalizedNumber extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'String';
        this.parameterTypes = [engine.numberType()];
        this.returnType = engine.stringType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return parameters[0].pipe(
            map( x => `${x}`)
        );
    }
}

export class ShortLocalizedNumber extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'ShortString';
        this.parameterTypes = [engine.numberType()];
        this.returnType = engine.stringType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return parameters[0].pipe(
            map( x => {
                if (x > 1000000) {
                    return `${Math.round(x / 100000) / 10} M`
                }
                else if (x > 1000) {
                    return `${Math.round(x / 100) / 10} K`
                }
                return `${x}`;
            })
        );
    }
}

export class LocalizedString extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'String';
        this.parameterTypes = [engine.stringType()];
        this.returnType = engine.stringType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return combineLatest(parameters as Observable<string>[]).pipe(
            map(params => {
                let key = params[0];
                for (let i = params.length - 1; i >= 1; --i) {
                    key = key.replace(`$${i}`, params[i]);
                }

                return key;
            })
        );
    }
}

export class Image extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'Image';
        this.parameterTypes = [engine.stringType()];
        this.returnType = engine.imageType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return parameters[0].pipe(
            map(x => new ImageContainer(x))
        );
    }
}

export class FontFamily extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'Font';
        this.parameterTypes = [engine.stringType(), engine.numberType()];
        this.returnType = engine.fontType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return combineLatest(parameters as Observable<any>[]).pipe(
            map(params => new FontContainer(params[0], params[1], null))
        );
    }
}

export class FontTyped extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'Font';
        this.parameterTypes = [engine.type('FontType')!, engine.numberType()];
        this.returnType = engine.fontType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return combineLatest(parameters as Observable<any>[]).pipe(
            map(params => new FontContainer(null, params[1], params[0]))
        );
    }
}

export class FontSized extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'Font';
        this.parameterTypes = [engine.numberType()];
        this.returnType = engine.fontType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return parameters[0].pipe(
            map(params => new FontContainer(null, params, null))
        );
    }
}

export class ColorAlpha extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'Color';
        this.parameterTypes = [engine.colorType(), engine.numberType()];
        this.returnType = engine.colorType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return combineLatest(parameters as Observable<any>[]).pipe(
            map(params => {
                const original = params[0] as ColorContainer;
                const alpha = params[1] as number;
                return new ColorContainer(original.red, original.green, original.blue, Math.max(0, Math.min(alpha, 1)));
            })
        );
    }
}