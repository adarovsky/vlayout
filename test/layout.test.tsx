import { mount } from 'enzyme';
import { Engine, Layout, pauseObserving, resizeObserver, resumeObserving } from '../src';
import React from 'react';
import { BehaviorSubject, EMPTY, of, Subscription, timer } from 'rxjs';
import { observers } from '../src/resize_sensor';

let module = require("../src/resize_sensor");
// const sizeChange = new Subject<ElementSize>();

let engine: Engine = new Engine();

beforeEach(() => {
    engine = new Engine();
});

describe("layout", () => {
    it("should remove all size observers on unmount", async function() {
        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
     layout {
         layer {
             stack {
                 center { x: 0.5 y: 0.5 }
                 label {
                     id: "image1"                     
                     text: "asd"                 
                 }
             }
         }
     }`}
            />
        );
        const node = wrapper.find(".vlayout_label_shadow");

        expect(observers).toHaveLength(4);
        expect(node.getDOMNode().childElementCount).toBe(3);
        pauseObserving();
        expect(node.getDOMNode().childElementCount).toBe(1);
        resumeObserving();
        expect(observers).toHaveLength(4);
        expect(node.getDOMNode().childElementCount).toBe(3);

        await timer(0).toPromise();
        wrapper.unmount();

        expect(observers).toHaveLength(0);
        resumeObserving();
        expect(observers).toHaveLength(0);
    });

    it("should not duplicate observers", async function() {
        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
     layout {
         layer {
             stack {
                 center { x: 0.5 y: 0.5 }
                 label {
                     id: "image1"
                     
                     text: "asd"                 
                 }
             }
         }
     }`}
            />
        );

        const node = wrapper.find(".vlayout_label_shadow");
        expect(node.getDOMNode().childElementCount).toBe(3);

        const observer = resizeObserver(node.getDOMNode());
        const subscription = new Subscription();

        subscription.add(observer.subscribe(() => {}));
        expect(node.getDOMNode().childElementCount).toBe(3);

        subscription.add(observer.subscribe(() => {}));
        expect(node.getDOMNode().childElementCount).toBe(3);

        wrapper.unmount();
        subscription.unsubscribe();
        await timer(50).toPromise();

        expect(observers).toHaveLength(0);
    });

    it("should not duplicate observers when toggle alpha", async function() {
        engine.registerButton('button1', async () => {});
        engine.registerButton('button2', async () => {});
        const test1 = new BehaviorSubject(0);
        engine.registerInput('test1', engine.numberType(), test1);
        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
     inputs {
         test1: Number
     }        
     layout {
         layer {
             vertical {
                 center { x: 0.5 y: 0.5 }
                 button1 {
                     id: "button1"
                     alpha: test1 == 0 ? 1 : 0
                     text: "asd1"                 
                 }
                 button2 {
                     id: "button1"
                     alpha: test1 == 1 ? 1 : 0
                     text: "asd2"                 
                 }
             }
         }
     }`}
            />
        );

        let node = wrapper.find(".vlayout_button");
        expect(node.getDOMNode().childElementCount).toBe(3);
        expect(node.text()).toBe('asd1');
        test1.next(1);
        wrapper.update();

        node = wrapper.find(".vlayout_button");
        expect(node.getDOMNode().childElementCount).toBe(3);
        expect(node.text()).toBe('asd2');
        wrapper.unmount();
        await timer(50).toPromise();

        expect(observers).toHaveLength(0);
    });

    it("should center view when paddings explicitly set to nil", async function() {
        const test1 = new BehaviorSubject(0);
        engine.registerInput('test', engine.numberType(), test1);
        const wrapper = mount(
            <Layout
                engine={engine}
                content={`
                inputs {
                    test: Number
                }        
                layout {
                    layer {
                        roundRect {
                            padding {
                                left: test % 2 == 0 ? 20 : nil
                                right: test % 2 == 0 ? 20 : nil
                                top: test % 2 == 0 ? 20 : nil
                                bottom: test % 2 == 0 ? 20 : nil
                            }
                            center {
                                x : test % 2 == 0 ? nil : 0.5
                                y : test % 2 == 0 ? nil : 0.5
                            }
                            fixedSize {
                                width: test % 2 == 0 ? nil : 120
                                height: test % 2 == 0 ? nil : 80
                            }
                            backgroundColor: #ffcccc
                        }
                    }
                }`}
            />
        );

        let node = wrapper.find(".vlayout_roundRect");
        wrapper.update();
        await timer(10).toPromise();
        expect(node.getDOMNode()).toMatchSnapshot();

        test1.next(1);
        await timer(10).toPromise();
        expect(node.getDOMNode()).toMatchSnapshot();
    });

    it("should pass classname from top level component to real layout element", async function() {
        const wrapper = mount(
          <Layout
            className={'sample-class'}
            engine={engine}
            content={`
                layout {
                }`}
          />
        );

        let node = wrapper.find(".vlayout_layout");
        expect(node.hasClass('sample-class')).toBe(true);
    });

    it("top level layout should not have min width and height", async function() {
        module.resizeObserver = jest.fn((e: Element) => {
            if (e.classList.contains('vlayout_label'))
                return of({width: 500, height: 300});
            return EMPTY;
        });

        const wrapper = mount(
            <Layout
                className={'sample-class'}
                engine={engine}
                content={`
                layout {
                    layer {
                        label {
                            padding { left: 0 right: 0 top: 0 bottom: 0 }
                            text: "test"
                        }
                    }
                }`}
            />
        );
        wrapper.update();

        let node = wrapper.find(".vlayout_layout");

        expect(node.prop('style')?.minWidth).toBeUndefined();
        expect(node.prop('style')?.minHeight).toBeUndefined();
    });
});
