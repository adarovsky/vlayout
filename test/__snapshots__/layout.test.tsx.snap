// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout should center view when paddings explicitly set to nil 1`] = `
<div
  class="vlayout_roundRect"
  style="position: absolute; left: 20px; z-index: 1; right: 20px; top: 20px; bottom: 20px; background-color: rgb(255, 204, 204); box-sizing: border-box;"
>
  <div
    style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden;"
  >
    <div
      style="position: absolute; left: 0px; top: 0px; width: 10000000px; height: 10000000px;"
    />
  </div>
  <div
    style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden;"
  >
    <div
      style="position: absolute; left: 0px; top: 0px; width: 200%; height: 200%;"
    />
  </div>
</div>
`;

exports[`layout should center view when paddings explicitly set to nil 2`] = `
<div
  class="vlayout_roundRect"
  style="position: absolute; z-index: 1; background-color: rgb(255, 204, 204); box-sizing: border-box; left: 50%; transform: translateX(-50%) translateY(-50%); top: 50%; max-width: 120px; min-width: 120px; width: 120px; max-height: 80px; min-height: 80px; height: 80px;"
>
  <div
    style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden;"
  >
    <div
      style="position: absolute; left: 0px; top: 0px; width: 10000000px; height: 10000000px;"
    />
  </div>
  <div
    style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden;"
  >
    <div
      style="position: absolute; left: 0px; top: 0px; width: 200%; height: 200%;"
    />
  </div>
</div>
`;
